### ISSUE TITLE
Support transformations for inter-grid kernels

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#134]
Issue #59 is adding support for inter-grid kernels to the Dynamo 0.3 API. However, to keep the size of that issue down, transformations are currently forbidden from being applied to loops containing inter-grid kernels. In this issue we will lift that restriction.

**Comment by arporter** [comment]
I've started extending the developers' guide to describe inter-grid kernels and the implications for halo exchanges. This will the lead on to the implications for transformations.

**Comment by arporter** [comment]
The available documentation on colouring and multi-grid is here: https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure/MeshColouring

**Comment by arporter** [comment]
I've stolen a couple of pictures for the Developers' Guide from Steve's LFRic presentation to the UM Partners Workshop in June. As of 1.3+ Sphinx has `:numref:` for referring to figure numbers so I've used that. I also had to add `numfig = True` to `conf.py`.

**Comment by arporter** [comment]
Currently, `DynKern.gen_code()` generates the code for the colour map look-up (and inserts it before the parent loop). It's done this way because this is also the place that has to insert the correct names for the various kernel arguments and that involves the colour map. However colour maps are associated with a mesh (in LFRic). We will therefore potentially need more than one (for invokes with multiple inter-grid kernels) *and* these will no longer be associated with the kernel argument that is being updated but with the argument on the coarse mesh.

**Comment by arporter** [comment]
The `DynMeshes` class stores a list of objects (dictionaries) describing the mesh information required for each kernel. I've begun to extend this with information on the colour-maps. I need to work out how to access this information from within `DynKern.gen_code()`. I must already do this for the argument names for inter-grid kernels...

**Comment by arporter** [comment]
@stevemullerworth: currently PSyclone generates different colour-map lookups depending on whether or not distributed memory is enabled. (We call `get_colour_map()` for DM and `get_colours(...)` if not.) Do we still need to do this or can we just always call the one colouring API now?

**Comment by stevemullerworth** [comment]
> @stevemullerworth: currently PSyclone generates different colour-map lookups depending on whether or not distributed memory is enabled. (We call `get_colour_map()` for DM and `get_colours(...)` if not.) Do we still need to do this or can we just always call the one colouring API now?

Firstly, the MPI API works even when running with no MPI because the get_last_halo_cell and get_last_edge_cell calls return sensible values (i.e the same values as you'd get from the ncp_colour array returned by get_colours). I suspect all our jobs are MPI-enabled, but many are run on a single MPI rank, so we know this is well-tested.


However, we don't have a truly non-MPI API in the mesh any more! Possibly we need "get_last_cell_per_colour" which would be easy to add to LFRic.

That would mean the PSy code for MPI would be:

```
      cmap => mesh%get_colour_map()
      !
      DO colour=1,mesh%get_ncolours()
        !$omp parallel default(shared), private(cell)
        !$omp do schedule(static)
        DO cell=1,mesh%get_last_XXXX_cell_per_colour(colour,Y)
```

where `XXXX` is "halo" for horizontally continuous or "edge" for discontinuos, and `Y` is halo depth (currently 1 but I guess will vary)

The PSy code for non-MPI would be:

```
      cmap => mesh%get_colour_map()
      !
      DO colour=1,mesh%get_ncolours()
        !$omp parallel default(shared), private(cell)
        !$omp do schedule(static)
        DO cell=1,mesh%get_last_cell_per_colour(colour)
```

as you don't need to worry about halos.

On the one hand, I guess this makes it easier to do transformations on several kernels some of which work on continuous and some on discontinuous spaces. On the other hand, it seems unlikely that we will be wanting to optimise non-MPI code so this might add complexity to the PSyclone API for no benefit.

So in summary, depending on your preference you can keep the same API for with/without MPI or you can adopt the latter API (and we will add the `get_last_cell_per_colour` function to LFRic).

**Comment by stevemullerworth** [comment]
PS. I got a bit ahead of myself. The old `get_colours` API does still exist (and can be accessed from either the mesh or the function space), but I suggest using the new one for consistency.

**Comment by arporter** [comment]
Thanks @stevemullerworth, that helps a lot. I'll use the new one as you suggest. From what you say I think I can just use `get_last_edge_cell_per_colour` if we're not doing DM - no need to add a new `get_last_cell_per_colour` unless you really want to :-)

**Comment by arporter** [comment]
The instance of `DynMeshes` is created by the constructor of `DynInvoke`. This happens during the construction of the Schedule and before any transformations are applied. Since colouring can only be added by a transformation, I can't decide which colour maps are required from within the `DynMeshes` constructor (since, at that point, we have no coloured loops).

**Comment by rupertford** [comment]
PR #217 (adding support for colouring intergrid kernels) has been merged to master. Keeping this issue open as other transformations will need to support intergrid kernels e.g. openmp.

**Comment by arporter** [comment]
Created `intergrid_omp` branch for adding OpenMP support.

**Comment by arporter** [comment]
If I simply turn off the check that prevents us doing OpenMP for inter-grid then I get, e.g.:

          DO colour=1,ncolour_fld_c
            !$omp parallel do default(shared), private(cell), schedule(static)
            DO cell=1,mesh_fld_c%get_last_halo_cell_per_colour(colour,1)
              !
              CALL restrict_kernel_code(nlayers, cell_map_fld_c(:,cmap_fld_c(colour, cell)),  &
                                                         ncpc_fld_m_fld_c, ncell_fld_m, fld_c_proxy%data,    &
                                                         fld_m_proxy%data, undf_any_space_1_fld_c,           &
                                                         map_any_space_1_fld_c(:,cmap_fld_c(colour, cell)),&
                                                         ndf_any_space_2_fld_m, undf_any_space_2_fld_m,&
                                                         map_any_space_2_fld_m)
            END DO 
            !$omp end parallel do
          END DO 

which looks OK. Note that `mesh_fld_c%get_last_halo_cell_per_colour(colour,1)` assumes we're using a depth-1 halo. Since we do not yet support redundant computation for inter-grid kernels this will always be True. @christophermaynard - does this look OK to you or have I missed something?

**Comment by arporter** [comment]
#236 was merged to master in December. Closing Issue.
